lib/rubygems/version.rb

```diff
@@ -388,51 +392,29 @@ def <=>(other)
     0
   end
 
+  # remove trailing zeros segments before first letter or at the end of the version
   def canonical_segments
     @canonical_segments ||= begin
-      numeric_segments, string_segments = _split_segments
```

Comment:
If this is too aggressive, there is an intermediate solution where we just inline the split segments code here which saves an array for the return value and sometimes avoids slicing entirely.